arch/x86/pagetables.c

```diff
@@ -246,28 +249,185 @@ void *vmap_user(void *va, mfn_t mfn, unsigned int order,
+    return _vunmap(&cr3, va);
+}
+
+int vunmap_user(void *va) {
+    return _vunmap(&user_cr3, va);
+}
+
+#define MMAP_RANGE_SIZE(size, mfn, flags, mmap_flags)                                    \
+    do {                                                                                 \
+        if (likely((mmap_flags) &MMAP_KERNEL))                                           \
+            kmap_##size((mfn), (flags));                                                 \
+        if ((mmap_flags) &MMAP_IDENT)                                                    \
+            vmap_##size(mfn_to_virt((mfn)), (mfn), (flags));                             \
+        if ((mmap_flags) &MMAP_USER)                                                     \
+            vmap_user_##size(mfn_to_virt_user((mfn)), (mfn), (flags | _PAGE_USER));      \
+        if ((mmap_flags) &MMAP_USER_KERNEL)                                              \
+            vmap_##size(mfn_to_virt_user((mfn)), (mfn), (flags & ~_PAGE_USER));          \
+        if ((mmap_flags) &MMAP_KERNEL_MAP)                                               \
+            vmap_##size(mfn_to_virt_map((mfn)), (mfn), (flags));                         \
+    } while (0)
+
+#define MMAP_RANGE_4K(mfn, flags, mmap_flags) MMAP_RANGE_SIZE(4k, mfn, flags, mmap_flags)
+#define MMAP_RANGE_2M(mfn, flags, mmap_flags) MMAP_RANGE_SIZE(2m, mfn, flags, mmap_flags)
+#define MMAP_RANGE_1G(mfn, flags, mmap_flags) MMAP_RANGE_SIZE(1g, mfn, flags, mmap_flags)
+
+void *mmap_range(paddr_t paddr, size_t size, unsigned long flags,
+                 mmap_flags_t mmap_flags) {
+    paddr_t cur = paddr;
+    paddr_t end = cur + size;
+
+    /* Round up to the next page boundary unless it is page aligned */
+    end = (end + PAGE_SIZE - 1) & PAGE_MASK;
+
+    spin_lock(&mmap_lock);
+
+    for (; (cur % PAGE_SIZE_2M) && cur < end; cur += PAGE_SIZE)
+        MMAP_RANGE_4K(paddr_to_mfn(cur), flags, mmap_flags);
+
+    if (likely(cur >= end))
+        goto done_4k;
+
+    for (; (cur % PAGE_SIZE_1G) && cur + PAGE_SIZE_2M <= end; cur += PAGE_SIZE_2M)
+        MMAP_RANGE_2M(paddr_to_mfn(cur), flags, mmap_flags);
+
+    if (likely(cur >= end))
+        goto done_2m;
+
+    for (; cur + PAGE_SIZE_1G <= end; cur += PAGE_SIZE_1G)
+        MMAP_RANGE_1G(paddr_to_mfn(cur), flags, mmap_flags);
+
+    for (; cur + PAGE_SIZE_2M <= end; cur += PAGE_SIZE_2M)
+        MMAP_RANGE_2M(paddr_to_mfn(cur), flags, mmap_flags);
+
+done_2m:
+    for (; cur < end; cur += PAGE_SIZE)
+        MMAP_RANGE_4K(paddr_to_mfn(cur), flags, mmap_flags);
+
+done_4k:
+    ASSERT(paddr_to_mfn(cur) == paddr_to_mfn(end));
+    spin_unlock(&mmap_lock);
+
+    if (mmap_flags & MMAP_KERNEL_MAP)
+        return paddr_to_virt_map(paddr);
+    else if (mmap_flags & MMAP_KERNEL)
+        return paddr_to_virt_kern(paddr);
+    else if (mmap_flags & MMAP_USER)
+        return paddr_to_virt_user(paddr);
+    else if (mmap_flags & MMAP_IDENT)
+        return paddr_to_virt(paddr);
```

Comment:
Ahh, here it's an if-else-if ladder, so should probably be one in `MMAP_RANGE_SIZE()` too.

This one is missing `MMAP_USER_KERNEL`, tho. So, maybe, using a `switch` block without a default label would be better, as it would warn about unhandled cases?:

```c
switch (mmap_flags) {
case MMAP_KERNEL_MAP:
    return paddr_to_virt_map(paddr);
case MMAP_KERNEL:
    return paddr_to_virt_kern(paddr);
case MMAP_USER:
    return paddr_to_virt_user(paddr);
case MMAP_IDENT:
    return paddr_to_virt(paddr);
}

return NULL;
```

However, the `mmap_flags` sanity check should probably be done at the begin of the function.

Reply:
Good point about the missing `MMAP_USER_KERNEL` and about turning it into a `switch` block. Will do.

But, the functionality I am trying to implement here is to be able to map given `mfn` to whatever memory region selected via the `mmap_flags`. This takes care of the KERNEL, IDENT and USER regions for the kernel pagetable and for the USER and KERNEL (called USER_KERNEL) for the user pagetable.
So we want to be able to map all selected regions with a one call and swoop. 
However, what this function returns for multiple regions is somewhat pre-defined as an implicit contract. It is defined by this if-else ladder. Essentially, callers receive the "top-most" memory region `va` address and if they want to use other region `va` from the ones specified, they have to calculate it on their own.

Reply:
If `mmap_flags` can be actually multiple flags, using a `switch` block won't work.

Returning the "highest" virtual address is also kinda fuzzy. I mean, the callers should know what flags they passed and thereby what address they should expect to get. But still, feels a little fragile to me... But if it orks for you, I shouldn't complain!